Quickstart tutorial with Next.js (#8269)
edgedb/edgedb: docs/intro/quickstart/modeling.rst

.. _ref_quickstart_modeling:

=================
Modeling the data
=================

.. edb:split-section::

>   The flashcards application has a simple data model, but it's interesting enough to get a taste of many of the features of the |Gel| schema language. You have a ``Card`` type that describes a single flashcard, which for now contains two required string properties: ``front`` and ``back``. Each ``Card`` belongs to a ``Deck``, and there is an explicit ordering to the cards in a given deck.

Reviewer: `of many features` instead of `of many of the features`, or we can say sth along the lines: The flashcards application has a straightforward data model, yet it's engaging enough to showcase many features of the Gel schema language.

>   The flashcards application has a simple data model, but it's interesting enough to get a taste of many of the features of the |Gel| schema language. You have a ``Card`` type that describes a single flashcard, which for now contains two required string properties: ``front`` and ``back``. Each ``Card`` belongs to a ``Deck``, and there is an explicit ordering to the cards in a given deck.

Reviewer 2: It looks like there's no `order` in the TypeScript version

Author: Right: it's just the order of the array itself. That's why I didn't mention an `order` property, just said "there _is_ an explicit ordering". I'll maybe make this more clear that I mean arrays are ordered and we want to keep this order in the database.

>   Looking at the mock data, you can see this structure in the JSON.

Reviewer: maybe to mention here that we mean the file `deck-edgeql.json`?

Author: `decks.json` is where the data lives once you import the `deck-edgeql.json`, so I'd like to keep that rather than make the actions module more complicated by needing to check if it exists, etc.

> maybe to mention here that we mean the file deck-edgeql.json?

Yeah, that's a good call. Maybe show a bit of the data itself.

  .. code-block:: typescript

>     interface Card {

Reviewer: in the tutorial app we define `RawJSONCard` and `RawJSONDeck` in both `lib/models` and in the `app/model`, is that needed?

Author: Yeah, we don't need both, PRs welcome, but that's more of an implementation detail than something related to the quickstart text itself. We don't actually mention those model modules directly.

    }

.. edb:split-section::

  Starting with this simple model, add these types to the ``default.gel`` schema file. As you can see, the types closely mirror the JSON mock data.

  Also of note, the link between ``Card`` and ``Deck`` objects creates a "1-to-n" relationship, where each ``Deck`` object has a link to zero or more ``Card`` objects. When you query the ``Deck.cards`` link, the cards will be unordered, so the ``Card`` type needs an explicit ``order`` property to allow sorting them at query time.

  .. code-block:: sdl-diff
    :caption: dbschema/default.gel

      module default {
    +   type Card {
    +     required order: int64;
    +     required front: str;
    +     required back: str;
    +   };
    +
    +   type Deck {
    +     required name: str;
    +     description: str;
    +     multi cards: Card {
    +       constraint exclusive;
    +     };
    +   };
      };

.. edb:split-section::

  Congratulations! This first version of the data model's schema is *stored in a file on disk*. Now you need to signal the database to actually create types for ``Deck`` and ``Card`` in the database.

  To make |Gel| do that, you need to do two quick steps:

>   1. **Create a migration**: a file with a list of low-level instructions.

Reviewer: I'd explain this a bit more clearly: a file containing a set of low level instructions that define how the database schema should change. It records any additions, modifications, or deletions to your schema in a way that the database can understand.

and something along the lines:
**Apply the migration**: This executes the migration file, instructing Gel to implement the recorded changes in the database. Essentially, this step updates the database structure to match your defined schema, ensuring that the Deck and Card types are created and ready for use.

> .. _ref_quickstart_modeling:

Reviewer 2: What if we add a bit more here? Since it's relatively short, maybe we can also explain links here?

Author: Yeah, I think in the mock-data -> Gel data version of this modeling section, we can look at the structure of the mock data and talk more about how to map this structure onto the Gel schema language's data modeling concepts, like links, cardinality, sets vs arrays, etc.

=================
Modeling our data
=================

.. edb:split-section::

  Our flashcards application has a simple data model, but it's interesting enough to get a taste of many of the features of the Gel schema language. We have a ``Card`` type that describes an single flashcard, which for now contains two required string properties: ``front`` and ``back``. Each ``Card`` belongs to a ``Deck``, and there is an explicit ordering to the cards in a given deck.

  Starting with this simple model, let's express these types in the ``default.gel`` schema file.

  .. code-block:: sdl-diff
    :caption: dbschema/default.gel

      module default {
    +   type Deck {
    +     required name: str;
    +     description: str;
    +   };

    +   type Card {

>     +     required order: int64;

Reviewer 2: I wonder if we should explain `order` a bit more - can be me, but I only got it after reading the next chapters 😅

Author: Yeah, good call, we can say that link sets are unordered, and since we want the cards to have a fixed order, we add this property so we can order them when we query for the deck and its cards.

    +     required front: str;
    +     required back: str;

    +     required deck: Deck;
    +   }
      };

.. edb:split-section::

>   Now that we've written our first version of our data model's schema, we will create a migration to apply this schema to the database. When making changes to our schema, the CLI migration tool will ask some questions to ensure that the changes we are making are what we expect. Since we had an empty schema before, the CLI will skip asking any questions and simply create the migration file.

Reviewer 3: > Now that we've written our first version of our data model's schema

This is a perfect opportunity to explain what a migration even means.

Basically:

```suggestion

  Congratulations! We have the first version of our data model's schema *stored in a file on disk*. Now we need to signal the database to actually create types for `Deck` and `Card` so you can store data and query data from them.
  
  To make Gel do that, we need to do two quick steps:
  
  * Create a migration: a file with a list of low-level instructions. 
  
    Note: when you are changing existing schema, the CLI migration tool might ask questions to ensure that it understands your changes exactly. 
  
  * Apply the migration: basically, tell Gel "I want you to use these instructions and get my types ready for action.
  
   Since we had an empty schema before, the CLI will skip asking any questions and simply create the migration file.
```

  .. code-block:: sh

      $ npx gel migration create
      Created ./dbschema/migrations/00001-m125ajr.edgeql, id: m125ajrbqp7ov36s7aniefxc376ofxdlketzspy4yddd3hrh4lxmla
      $ npx gel migrate
      Applying m125ajrbqp7ov36s7aniefxc376ofxdlketzspy4yddd3hrh4lxmla (00001-m125ajr.edgeql)
      ... parsed
      ... applied

.. edb:split-section::

>   Let's take a look at the schema we've generated in our built-in database UI. We can use this tool to visualize our data model and see the object types and links we've defined.

Reviewer 3: We need like a gallery widget here. First screenshot will be dashboard with circled button for "Schema Viewer", second will be the viewer

```suggestion
  Let's take a look at the schema we've generated in our built-in database UI. We can use it to visualize our data model and see the object types and links we've defined.
```
